source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,348 @@
+  }
+}
+
+Router::UpstreamRequest::UpstreamRequest(Router& parent, Tcp::ConnectionPool::Instance& pool,
+                                         MessageMetadataSharedPtr& metadata,
+                                         SerializationType serialization_type,
+                                         ProtocolType protocol_type)
+    : parent_(parent), conn_pool_(pool), metadata_(metadata),
+      deserializer_(
+          NamedDeserializerConfigFactory::getFactory(serialization_type).createDeserializer()),
+      protocol_(NamedProtocolConfigFactory::getFactory(protocol_type).createProtocol()),
+      request_complete_(false), response_started_(false), response_complete_(false) {}
+
+Router::UpstreamRequest::~UpstreamRequest() {}
+
+Network::FilterStatus Router::UpstreamRequest::start() {
+  Tcp::ConnectionPool::Cancellable* handle = conn_pool_.newConnection(*this);
+  if (handle) {
+    // Pause while we wait for a connection.
+    conn_pool_handle_ = handle;
+    return Network::FilterStatus::StopIteration;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Router::UpstreamRequest::resetStream() {
+  if (conn_pool_handle_) {
+    ASSERT(!conn_data_);
+    conn_pool_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
+    conn_pool_handle_ = nullptr;
+    ENVOY_LOG(debug, "dubbo upstream request: reset connection pool handler");
+  }
+
+  if (conn_data_) {
+    ASSERT(!conn_pool_handle_);
+    conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+    conn_data_.reset();
+    ENVOY_LOG(debug, "dubbo upstream request: reset connection data");
+  }
+}
+
+void Router::UpstreamRequest::encodeData(Buffer::Instance& data) {
+  ASSERT(conn_data_);
+  ASSERT(!conn_pool_handle_);
+
+  ENVOY_STREAM_LOG(trace, "proxying {} bytes", *parent_.callbacks_, data.length());
+  conn_data_->connection().write(data, false);
+}
+
+void Router::UpstreamRequest::onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                                            Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(warn, "dubbo upstream request: connection failure '{}'", host->address()->asString());
+
+  conn_pool_handle_ = nullptr;
+
+  // Mimic an upstream reset.
+  onUpstreamHostSelected(host);
+  onResetStream(reason);
+
+  parent_.upstream_request_buffer_.drain(parent_.upstream_request_buffer_.length());
+
+  // If it is a connection error, it means that the connection pool returned
+  // the error asynchronously and the upper layer needs to be notified to continue decoding.
+  // If it is a non-connection error, it is returned synchronously from the connection pool
+  // and is still in the callback at the current Filter, nothing to do.
+  if (reason == Tcp::ConnectionPool::PoolFailureReason::Timeout ||
+      reason == Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure ||
+      reason == Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure) {
+    parent_.callbacks_->continueDecoding();
```

Comment:
Other PoolFailureReason do not require continueDecoding?

Reply:
Is not needed, if the reason for failure is equal to overflow, the newConnection call will return nullptr, UpstreamRequest: :start call will return to Continue.

Reply:
```
if (reason != Tcp::ConnectionPool::PoolFailureReason::Overflow) {
   parent_.callbacks_->continueDecoding();
}
```
better?

Reply:
Haha, I wrote it like this at first, but later I changed it because I thought it would be easier to understand.